File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if (root == null) return false;

        if (checkTree(root, subRoot)) return true;

        return isSubtree(root.left, subRoot) ||
                isSubtree(root.right, subRoot);
    }

    private boolean checkTree(TreeNode node, TreeNode subNode) {
        if (node == null || subNode == null) {
            return node == null && subNode == null;
        }

        if (node.val != subNode.val) return false;
        return checkTree(node.left, subNode.left) &&
                checkTree(node.right, subNode.right);
    }
}

File: 0572-subtree-of-another-tree/0572-subtree-of-another-tree.py
class Solution:
    def isSubtree(self, root, subRoot):
        if not root:
            return False

        if self.sameTree(root, subRoot):
            return True

        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def sameTree(self, root, subRoot):
        if not root or not subRoot:
            return not root and not subRoot

        if root.val != subRoot.val:
            return False

        return self.sameTree(root.left, subRoot.left) and self.sameTree(root.right, subRoot.right)
